### ISSUE TITLE
0 val mAP

### ISSUE THREAD

**Issue by Vincent9797** [fizyr/keras-retinanet#1343]
As seen in the image below, training loss decreases. However, val regression loss remains at 0 and val mAP is always 0 for all classes.

![image](https://user-images.githubusercontent.com/42912553/80387223-fa313b80-88da-11ea-8e53-dd559d4180f2.png)

Command to run training: 
`python keras_retinanet/bin/train.py --batch-size 16 --image-min-side 40 --image-max-side 160 --compute-val-loss csv bin_images/train_with_1k_filtered.csv bin_images/classes.csv --val-annotations bin_images/test_labels.csv`

Images are prepared with labelImg. Images are binary and look like the one below:
![image](https://user-images.githubusercontent.com/42912553/80387555-5ac07880-88db-11ea-9ec1-05b4c4fcbadd.png)

**Comment by hgaiser** [comment]
Could you try with debug.py to see if your annotations are "covered" by anchors?

**Comment by Vincent9797** [comment]
Thanks for your help! `debug.py` showed that the annotations were not covered by the anchors generated. After changing the anchor parameters it works.

**Comment by ChunmianLin** [comment]
> Thanks for your help! `debug.py` showed that the annotations were not covered by the anchors generated. After changing the anchor parameters it works.

Hi, I encountered the same problem that **image doesn't covered by annotated bbox when I run demo.py**, and **detection mAP is nearly zero.** 
How do you solve this problem? Could you please help me?
Thanks a lot !!!

**Comment by ChunmianLin** [comment]
Also, my training loss is abnormal. It starts a relative small value (about 3.5-4.0), and converges very quickly after one epoch (about 1.0-1.5 at epoch one). Then, it cannot change significantly, and classification loss metric is always small than 0.5.
Do you know what made this problem happen? I really have no idea!
My data mostly consists of dense and occlusion object in urban scene.
Thanks in advance!

**Comment by Vincent9797** [comment]
> Hi, I encountered the same problem that image doesn't covered by annotated bbox when I run demo.py, and detection mAP is nearly zero.

Does this refer to images that are in your training/validation set? For my case, it was because I did not change the coordinates of my bounding boxes when I processed it. The scenario is similar to the 2 images attached. The generated Bbox overlaps with the GT in the training images. The same thing happens for the validation image.

 But the error was that the **Bbox in my validation image was not adjusted when I did certain augmentations to the original image.**
![train](https://user-images.githubusercontent.com/42912553/86904731-b14ffb80-c143-11ea-92e0-0975976545ce.jpg)
![val](https://user-images.githubusercontent.com/42912553/86904723-b01ece80-c143-11ea-82d6-002311ca947f.jpg)

**Comment by ChunmianLin** [comment]
> For my case, it was because I did not change the coordinates of my bounding boxes when I processed it. The scenario is similar to the 2 images attached. The generated Bbox overlaps with the GT in the training images. The same thing happens for the validation image.
> 
> But the error was that the **Bbox in my validation image was not adjusted when I did certain augmentations to the original image.**
> ![train](https://user-images.githubusercontent.com/42912553/86904731-b14ffb80-c143-11ea-92e0-0975976545ce.jpg)
> ![val](https://user-images.githubusercontent.com/42912553/86904723-b01ece80-c143-11ea-82d6-002311ca947f.jpg)
> 
> For your case, I can;t be sure of the error unless you show me some examples Could you show me some images of the output generated by `demo.py`?

Hi, I cannot understand "**it was because I did not change the coordinates of my bounding boxes when I processed it. The scenario is similar to the 2 images attached. The generated Bbox overlaps with the GT in the training images. The same thing happens for the validation image.** " So sorry for my ignorance. Could you please explain for me in detail? I just start with this project few days. 
My problems have two points:
(1) debug.py: Though image can show normally, no any GT annotation covers  the corresponding object. What cause this bug?
(2) training loss: my model training loss is quite small and converges very fast in the first epoch. But after that, insignificant value fluctuation is observed. I guess, there is useless training process and no meaningful information can be learned by the model?

Thanks for your response! And I would appreciate you a lot if any help you provide!

**Comment by Vincent9797** [comment]
The 2 images are generated by demo.py. Basically the top image is a RGB image taken from a camera feed, and the bottom image was after I processed the image. When I processed it, it was resized, while the original bbox coordinates was not changed. As such, you can see that the GT boxes have the correct position relative to each other. That was why my validation MAP was so low - my Bbox coordinates were wrong after I processed the RGB image.

1. The main issue that causes this is the parameters for the anchors generated. The image below is from keras-retinanet/keras_retinanet/utils/anchors.py. It shows that the original anchors are generated with certain parameters such as ratio. These sometimes need to be adjusted to suit your use case.

![image](https://user-images.githubusercontent.com/42912553/86946372-d90f8580-c17c-11ea-8a3d-f353c62460ce.png)

2. I can't conclude whether the model has stopped learning after the first epoch. The difficulty in learning could be due to your GT annotations not being covered by the anchors in the first place. So what I would do is to make sure your GT is covered by the anchors before even starting with the training.
